### ISSUE TITLE
Further progress on CuPy wrapper

### ISSUE THREAD

**Issue by ericmjl** [HIPS/autograd#406]
Looping back on wrapping CuPy. 

Summary:

- Can do gradient descent-based optimization with CuPy wrapper for a simple linear model.
- Tests for wrapped functions using `check_grad` are now annotated.

(1) Gradient-based optimization

[This](https://github.com/ericmjl/autograd/blob/cupy/tests/test_cupy_hack.py#L52) now works for a simple linear model that I coded up. 

(2) Checking `check_grad` tests

[`test_cupy.py`](https://github.com/ericmjl/autograd/blob/cupy/tests/test_cupy.py) is a direct mirror of [`test_numpy.py`](https://github.com/ericmjl/autograd/blob/cupy/tests/test_numpy.py), except with pytest annotations giving me a hint as to why they might be failing. 

For some, I have an idea on how to debug (related to minute differences between CuPy and NumPy), for others, I might start reaching out for advice on what's going on. 

cc'ing people who I remember are interested, and those with whom I've conferred:

@mattjj, @dougalm, @duvenaud, @jermwatt, @kmaehashi, @dwd31415, @tscohen, @oeway, @iaroslav-ai, @bingo619, @kastnerkyle, @outlace, @samuela

**Comment by ericmjl** [comment]
More progress today! In [the tests for CuPy in my branch](https://github.com/ericmjl/autograd/blob/cupy/tests/test_cupy.py), I have 42 failed and 57 passed tests. Officially >5/9 of the way through! Of the tests that passed, they were mostly due to minor differences between the CuPy and NumPy APIs. 

Along the way, I have also run the `test_numpy.py` suite to make sure nothing there breaks.

Finally, I have made some changes that take advantage of CuPy to automagically distinguish between the two array libraries. This is hacky, but it works. I might try to make this more generalized in the future.

**Comment by ericmjl** [comment]
More progress today! I have knocked out about 2/3 of the tests. Since >1 week ago, there's been enough passing tests coverage to write plain vanilla (Feed Forward) neural networks using nothing but dot products and loss functions. 

My next most immediately useful thing would be to wrap the CuPy sparse library, which is what I'll be moving forward on.

**Comment by brandondube** [comment]
How’s the progress on this these days?

**Comment by ericmjl** [comment]
http://github.com/ericmjl/autograd-cupy

_That said_... I would suggest using Jax instead! 

My autograd CuPy wrapper was very educational, but I'm ultimately not in a position to maintain it long-term.
